File: touchfish-core/src/domain/entity/topic.rs
use strum_macros::{Display, EnumString};
use yfunc_rust::{YTime, prelude::*};
use serde::{Deserialize, Serialize};

#[yfunc]
#[derive(Serialize, Debug)]
pub struct Topic {
    pub uid: String,
    pub topic_type: TopicType,
    pub subject: String,
    pub title: String,
    pub messages: Vec<Message>,
    pub extra_info: TopicExtraInfo,
    pub create_time: YTime,
    pub update_time: YTime,
}

#[yfunc]
#[derive(Serialize, Deserialize, Debug, EnumString, Display)]
pub enum TopicType {
    Info, Warning, Error,
}

#[yfunc]
#[derive(Serialize, Deserialize, Debug)]
pub struct TopicExtraInfo {

}

#[yfunc]
#[derive(Serialize, Debug)]
pub struct Message {
    pub uid: String,
    pub level: MessageLevel,
    pub source: String,
    pub title: String,
    pub body: String,
    pub has_read: bool,
    pub extra_info: MessageExtraInfo,
    pub create_time: YTime,
    pub update_time: YTime,
}

#[yfunc]
#[derive(Serialize, Debug)]
pub enum MessageLevel {
    Info, Warning, Error,
}

#[yfunc]
#[derive(Serialize, Debug)]
pub struct MessageExtraInfo {

}

File: touchfish-macos-client/TouchFish/Topic/Topic.swift
import AppKit

struct Topic {
    
    static var unreadMsgCount = 0
    
    let uid: String
    let subject: String
    let source: String
    let title: String
    let messages: [Message]
    let extraInfo: [String:String]
    let createTime: String
    let updateTime: String
    
    var infoCount: Int {
        return messages.filter { $0.level == .Info }.count
    }
    
    var warningCount: Int {
        return messages.filter { $0.level == .Warning }.count
    }
    
    var errorCount: Int {
        return messages.filter { $0.level == .Error }.count
    }
    
    var unreadCount: Int {
        return messages.filter { !$0.hasRead }.count
    }
    
}

struct Message {
    
    enum Level: String {
        case Info
        case Warning
        case Error
    }
    
    struct ExtraInfo: Codable {
        
    }
    
    let uid: String
    let level: Level
    let title: String
    let body: String
    let hasRead: Bool
    let extraInfo: [String:String]
    let createTime: String
    let updateTime: String
    
}

